### PR TITLE
LLM Forward Step

### PR DESCRIPTION
**Pull request by maanug-nv**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/tron/train.py [hunk 1]
```diff
@@ -424,10 +425,25 @@ def train_step(
...
         optimizer.zero_grad()
 
+        # Optionally inject state into forward step
+        num_fw_args = len(inspect.signature(forward_step_func).parameters)
+        fail_msg = f"""
+        forward_step_func has {num_fw_args} arguments. Only the following signatures are supported: 
+            2 args: forward_step_func(data_iterator: Iterable, model: GPTModel)
+            3 args: forward_step_func(state: GlobalState, data_iterator: Iterable, model: GPTModel)
+        """
+        assert num_fw_args in [2, 3], fail_msg
```
Comment (ananthsub) [inline nemo/tron/train.py:435]:
```suggestion
        assert num_fw_args in (2, 3), fail_msg
```

File: nemo/tron/train.py [hunk 2]
```diff
@@ -424,10 +425,25 @@ def train_step(
...
+
+        if num_fw_args == 3:
+            # inject global_state
+            wrapped_forward_step = partial(forward_step_func, global_state)
+        else:
+            wrapped_forward_step = forward_step_func
```
Comment (ananthsub) [inline nemo/tron/train.py:441]: you can check the number of parameters outside of the loop so we only do it once. only the partial needs to be done in the loop

Comment (maanug-nv) [reply nemo/tron/train.py]: not sure if you mean just outside the rerun loop or something else. but actually, i think it makes sense to do this before even calling train().

Comment (maanug-nv) [reply nemo/tron/train.py]: Updated. let me know if you think there might be any issues with it

File: nemo/tron/api.py [hunk 3]
```diff
@@ -39,6 +40,15 @@ def megatron_pretrain(
...
     ckpt_context = setup_output.checkpointing_context
 
+    # Check num args to forward_step_func
+    num_fw_args = len(inspect.signature(forward_step_func).parameters)
+    fail_msg = f"""
+    forward_step_func has {num_fw_args} arguments. Only the following signatures are supported: 
+        2 args: forward_step_func(data_iterator: Iterable, model: GPTModel)
+        3 args: forward_step_func(state: GlobalState, data_iterator: Iterable, model: GPTModel)
+    """
+    assert num_fw_args in (2, 3), fail_msg
```
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/tron/api.py:50]: Just noting here that this arg validation is great, but the arg requirements should be included in the docstring (which doesn't exist). We should add those in a separate PR.

File: nemo/tron/config.py [hunk 4]
```diff
@@ -115,6 +115,9 @@ class GPTDatasetConfig(MCoreGPTDatasetConfig):
     data_sharding: bool = True
     """Disable data sharding."""
 
+    create_attention_mask_in_dataloader: bool = True
+    """If set, do not create attention_masks in dataloader."""
```
Comment (ericharper) [inline nemo/tron/config.py:119]: What happens if flash attention is being used and this is set to True?

Comment (maanug-nv) [reply nemo/tron/config.py:119]: looks like this is a duplicate of something that already exists: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/gpt_dataset.py#L38-L40

MLM args just use a different name: https://github.com/NVIDIA/Megatron-LM/blob/main/pretrain_gpt.py#L284

**Review by ananthsub — APPROVED** [review]

File: nemo/tron/utils/train_utils.py [hunk 5]
```diff
@@ -509,3 +512,25 @@ def reduce_aux_losses_tracker_across_ranks():
             torch.distributed.all_reduce(values, group=tracker[name].get("reduce_group"))
         if tracker[name].get("avg_group") is not None:
             torch.distributed.all_reduce(values, group=tracker[name]["avg_group"], op=torch.distributed.ReduceOp.AVG)
+
+
+def maybe_inject_state(forward_step_func: Callable, state: GlobalState, num_fw_args: Optional[int] = None) -> Callable:
```
Comment (hemildesai) [inline nemo/tron/utils/train_utils.py:517]: These are used in eval too. Should we move them to `common_utils.py`?
